### DISCUSSION TITLE
tck-maven-plugin

### DISCUSSION THREAD

**Discussion by laeubi** [osgi/osgi#548] [Ideas]
In the last spec-call there was some discussion about executing TCK tests and I think one important part would be to have an easy way for no-domain-experts to execute TCKs, this is especially important for things like where providers of a JDBC driver at best implement the [Data Service Specification for JDBC™ Technology](https://docs.osgi.org/specification/osgi.cmpn/8.0.0/service.jdbc.html)  as they are the domain experts for their driver but not domain experts of OSGi.
In such context, there is often a demand to have a minimal invasive support (what works e.g. with [bnd-process-goal](https://github.com/bndtools/bnd/tree/master/maven-plugins/bnd-maven-plugin#bnd-process-goal) to generate required headers) and at best they just can test their compatibility in a similar non-invasive way without additional knowledge of OSGi/TCK and tools (what often is a reason why OSGi support is dropped from libraries).

I therefore like to propose a  `tck-maven-plugin` that supports just one specific task, that is executing a TCK artifact from maven against some other artifacts from the project or the maven repository. Such a plugin can use BND internally or whatever technique suffice, but should not expose any of those to the user and therefore can be configured completely by maven standard techniques people are familiar with.

The proposed configuration of such a plugin should looks like this:
```
<plugin>
	<groupId>org.osgi.maven</groupId>
	<artifactId>tck-plugin</artifactId>
	<version>....</version>
	<configuration>
		<framework>
			<!-- specify the maven artifact used for the osgi framework in the execution -->
			<dependency>
				<groupId>org.eclipse.platform</groupId>
				<artifactId>org.eclipse.osgi</artifactId>
				<version>....</version>
			</dependency>
		</framework>
		<tcks>
			<!-- specify the maven artifacts of one or more TCK used in the execution -->
			<dependency>
				<groupId>org.osgi</groupId>
				<artifactId>org.osgi.service.jdbc</artifactId>
				<version>....</version>
			</dependency>
		<tcks>
		<bundles>
			<!-- specify the maven artifacts of one or more bundles that should be run agains the TCK -->
			<dependency>
				<groupId>com.h2database</groupId>
				<artifactId>h2</artifactId>
				<version>....</version>
			</dependency>
			<dependency>	
				...
			</dependency>
		</bundles>
		<results>${project.build.directory}/tck-results</result>
		....
	</configuration>
</plugin>
```

I can create such maven-plugin as a PR for further investigation / discussions but first want to gather some feedback if such thing could be accepted or there are more ideas and/or concerns.

**Comment by bjhargrave** [comment]
This is just a bndrun file in xml form. Yuck. 

We should instead show how to use the existing bnd-testing-maven-plugin to run TCKs. I don't want OSGi to be in the business of building and supporting tooling. So -1 from me for having the OSGi project develop and maintain a maven plugin. 

And what about gradle users? Would we also have to develop and maintain a gradle tck plugin for them? No need since Bnd already has Gradle support for OSGi testing! :-)

**Reply by laeubi** [reply]
As said, any bnd specific configuration file is actually a no-go for me. Most open-source projects can simply not afford the effort learning another domain specific language, beside that it never has been "just" a bndrun file when I look at the "simple" examples, and it always duplicates configurations from the maven model in the bnd config so one has to take care it is in sync.

But if you can share an example (e.g. for the mentioned JDBC TCK and `h2` artifact) this might help to get a better understanding of what you propose.

Regarding gradle I can't offer any insights as most of the time maven is used as of today ...

**Reply by bjhargrave** [reply]
> Most open-source projects can simply not afford the effort learning another domain specific language,

So they can learn a maven plugin's configuration schema but not a properties schema (bndrun files are Properties files)?



> But if you can share an example

I will work on a TCK example using bnd-testing-maven-plugin.

**Reply by laeubi** [reply]
> So they can learn a maven plugin's configuration schema but not a properties schema (bndrun files are Properties files)?

Because they use "maven" for their build, and there are millions of maven plugins where they can use such things, but there is only one "bnd-run-roperties" file format full of bnd-specific things they do not work with in their usual bussiness.

**Comment by grgrzybek** [comment]
Let me add a comment as another person that just started using `bnd` tool despite using OSGi for almost 10 years with Felix' `maven-bundle-plugin`.

I followed [this guide|https://github.com/osgi/osgi/blob/main/CONTRIBUTING.md#ide-setup]:
```
Before importing the project into Eclipse, it is recommended to do a full build as described above. Then use Import... > Existing Projects into Workspace in Eclipse.
```

This part was easy - Eclipse discovered `bnd` workspace's projects:
![image](https://user-images.githubusercontent.com/306565/211248107-821ec992-9c03-409f-998d-c174f419a309.png)

And then simply irrecoverable (I couldn't even close the project):
![image](https://user-images.githubusercontent.com/306565/211248158-a4dc8a88-26ad-4ad3-b069-d2078910b903.png)

Then I tried the lower-level approach:
```
$ cat `which bnd`
#!/bin/bash

java -jar /home/ggrzybek/.m2/repository/biz/aQute/bnd/biz.aQute.bnd/6.4.0/biz.aQute.bnd-6.4.0.jar $@

$ pwd
/home/ggrzybek/sources/github.com/osgi/osgi

$ bnd build org.osgi.test.cases.servlet
-----------------
Errors
000: Exception: java.lang.Error: Unresolved compilation problems: 
	The import aQute cannot be resolved
	The import aQute cannot be resolved
	The import aQute cannot be resolved
	The import aQute cannot be resolved
	The import aQute cannot be resolved
	The import aQute cannot be resolved
	The import aQute cannot be resolved
	The import aQute cannot be resolved
	The import aQute cannot be resolved
	The import aQute cannot be resolved
	The import aQute cannot be resolved
	AnalyzerPlugin cannot be resolved to a type
	Analyzer cannot be resolved to a type
	ProjectBuilder cannot be resolved to a type
	Create cannot be resolved
	ProjectBuilder cannot be resolved to a type
	ProjectBuilder cannot be resolved to a type
...
```

I don't give up, I'm just analyzing https://github.com/osgi/osgi/actions/runs/3451110034/jobs/5760142348 and which commands are used in this GH action. The logs there are:
```
cd osgi.tck.cmpn
  java -jar jar/bnd.jar --exceptions runtests --title org.osgi.test.cases.servlet --reportdir reports/org.osgi.test.cases.servlet org.osgi.test.cases.servlet.bnd
  shell: /usr/bin/bash --noprofile --norc -e -o pipefail {0}
  env:
    LC_ALL: en_US.UTF-8
    GRADLE_OPTS: -Dorg.gradle.parallel=true -Dmaven.repo.local=cnf/generated/m2
    canonical: false
    java_version: 17
...
```

But I don't see `osgi.tck.cmpn` folder in osgi/osgi project... If I don't run single TCK on my laptop, I'll ask on Slack next week...

Summarizing - I love the abstract idea of workspace and projects, sharing the outcome via repositories and generally the model. But implementation of the ideas didn't give me the best possible impression (for now).

And yes - I used `gradle` to build everything possible, as specified in https://github.com/osgi/osgi/blob/main/CONTRIBUTING.md#build-environment.

**Reply by bjhargrave** [reply]
> But I don't see `osgi.tck.cmpn` folder in osgi/osgi project

The build on GHA uses some code to cause the TCK extraction:

https://github.com/osgi/osgi/blob/ec62218646e12d2cb8069a7f685b80a159f61805/osgi.tck/build.gradle#L105-L121

The task is `:osgi.tck:preptck.cmpn`. If you run that gradle task after the `:build` task, you will have a `osgi.tck/generated/osgi.tck.cmpn` folder that you can `cd` into the run the TCKs.

The GHA build uploads these folders

https://github.com/osgi/osgi/blob/ec62218646e12d2cb8069a7f685b80a159f61805/.github/workflows/cibuild.yml#L79-L86

which are then downloaded and used by the tck execution job.

https://github.com/osgi/osgi/blob/ec62218646e12d2cb8069a7f685b80a159f61805/.github/workflows/cibuild.yml#L191-L199

**Reply by grgrzybek** [reply]
ah - that's very helpful. Thanks!

**Reply by grgrzybek** [reply]
whew - I learned a lot about Github actions and `cibuild.yml` helped me to understand how powerful they are. Thanks!

**Comment by laeubi** [comment]
@grgrzybek I have added a TCK to the Microsoft JDBC driver here that shows how to do the maven setup:
- https://github.com/microsoft/mssql-jdbc/pull/2066/files

let me know if you need any help with the setup for pax-web, but it should be quite straight forward with this example.

**Reply by grgrzybek** [reply]
Thanks @laeubi for info - indeed this is much easier using Maven ;) I start PTO tomorrow and I have to finish some Keycloak related issues in Pax Web, so I'll get back to this next week.
